Conversion between pixel_mask and image_mask

mamelara opened the pull request:
## Motivation

Addresses Issue #669 

## How to test the behavior?

Unit tests are included. To run the tests, use tox.
## Checklist

- [X] Have you checked our [Contributing](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/docs/CONTRIBUTING.rst) document?
- [X] Have you ensured the PR description clearly describes problem and the solution?
- [X] Is your contribution compliant with our coding style ? This can be checked running `flake8` from the source directory.
- [X] Have you checked to ensure that there aren't other open [Pull Requests](https://github.com/NeurodataWithoutBorders/pynwb/pulls) for the same change?
- [X] Have you included the relevant issue number using `#XXX` notation where `XXX` is the issue number ?

mamelara committed:
Add methods that convert pixel to image and vice versa

Work in progress for creating methods that convert pixel to image and image to pixel mask. Eventually, will need to handle image masks that have multiple image matrices.

mamelara committed:
Add tests for masks conversion

mamelara commented:
This is still WIP, but just wanted to see if I'm going in the right direction with this. I also need to take into account that an image_mask can have multiple image matrices in it.

oruebel commented on src/pynwb/ophys.py:
```diff
@@ -271,6 +271,27 @@ def add_roi(self, **kwargs):
 
+    def pixel_to_image(self, pixel_mask, width, height):
+        image_matrix = np.zeros((width, height))
+        for mask in pixel_mask:
```
I think something like this might be a bit faster than setting all the elements one-by-one in the for loop:

```
npmask = np.asarray(pixel_mask)
image_matrix[npmask[:,0], npmask[:,1]] =  npmask[:,-1]
return image_matrix
```

oruebel commented on src/pynwb/ophys.py:
```diff
@@ -271,6 +271,27 @@ def add_roi(self, **kwargs):
+
+    def image_to_pixel(self, image_mask):
+        """Convert a image_mask to a pixel_mask that contains all the elements of the form (x, y, weight)"""
+        width, height = np.shape(image_mask)
```
Here ``numpy nditer`` might help simplify things:

```
pixel_mask = []
it = np.nditer(image_mask, flags=['multi_index'])
while not it.finished:
    pixel_mask.append( [it.multi_index[0], it.multi_index[1], it[0][()]] )
    it.iternext()
return pixel_mask



```

mamelara committed:
Refactor conversion methods

Uses numpy's way of indexing and iterating over numpy arrays to create images and pixels.

mamelara commented:
Did some of the suggestions given. Only failing test is creating pixel_masks out of multiple images as given in this example in `test_init`:

```python
img_mask = [[[1.0 for x in range(w)] for y in range(h)], [[2.0 for x in range(w)] for y in range(h)]]
pix_mask = [[1, 2, 1.0], [3, 4, 1.0], [5, 6, 1.0],
                    [7, 8, 2.0], [9, 10, 2.0]]
``` 

For clarification, does the pixel_mask just treat the two w x h arrays as just one giant coordinate? Meaning that, pix_mask [7, 8, 2.0] refers to the x-y coordinates of the second image?

mamelara committed:
Rename tests and create helper method

Modify some of the tests and also add a helper function for creating a
basic plane segmentation when the boiler plate objects are not needed
for testing

mamelara committed:
Flake8 fixes

bendichter commented on src/pynwb/ophys.py:
```diff
@@ -272,6 +272,27 @@ def add_roi(self, **kwargs):
             rkwargs['voxel_mask'] = voxel_mask
         return super(PlaneSegmentation, self).add_row(**rkwargs)
 
+    def pixel_to_image(self, pixel_mask):
```
```suggestion
        @staticmethod
    def pixel_to_image(pixel_mask):
        image_matrix = np.zeros(np.shape(pixel_mask))
        npmask = np.asarray(pixel_mask)
        x_coords = npmask[:, 0].astype(np.int)
        y_coords = npmask[:, 1].astype(np.int)
        weights = npmask[:, -1]
        image_matrix[y_coords, x_coords] = weights
        return image_matrix

    @staticmethod
    def image_to_pixel(image_mask):
```

bendichter replied:
```suggestion
    def pixel_to_image(pixel_mask):
```

bendichter commented on src/pynwb/ophys.py:
```diff
@@ -272,6 +272,27 @@ def add_roi(self, **kwargs):
             rkwargs['voxel_mask'] = voxel_mask
         return super(PlaneSegmentation, self).add_row(**rkwargs)
 
```
```suggestion

    @staticmethod
```

bendichter commented on src/pynwb/ophys.py:
```diff
@@ -272,6 +272,27 @@ def add_roi(self, **kwargs):
+        weights = npmask[:, -1]
+        image_matrix[y_coords, x_coords] = weights
+        return image_matrix
+
```
```suggestion

    @staticmethod
```

bendichter commented on src/pynwb/ophys.py:
```diff
@@ -272,6 +272,27 @@ def add_roi(self, **kwargs):
+        image_matrix[y_coords, x_coords] = weights
+        return image_matrix
+
+    def image_to_pixel(self, image_mask):
```
```suggestion
    def image_to_pixel(image_mask):
```

bendichter commented on src/pynwb/ophys.py:
```diff
@@ -272,6 +272,27 @@ def add_roi(self, **kwargs):
+    def pixel_to_image(self, pixel_mask):
+        image_matrix = np.zeros(np.shape(pixel_mask))
+        npmask = np.asarray(pixel_mask)
+        x_coords = tuple(map(int, npmask[:, 0]))
```
```suggestion
        x_coords = npmask[:, 0].astype(np.int)
```

bendichter commented on src/pynwb/ophys.py:
```diff
@@ -272,6 +272,27 @@ def add_roi(self, **kwargs):
+        image_matrix = np.zeros(np.shape(pixel_mask))
+        npmask = np.asarray(pixel_mask)
+        x_coords = tuple(map(int, npmask[:, 0]))
+        y_coords = tuple(map(int, npmask[:, 1]))
```
```suggestion
        y_coords = npmask[:, 1].astype(np.int)
```

oruebel commented:
@mamelara what is the status of this PR? Looks like there were a few minor changes requested, could you add those so we can close and merge?

mamelara committed:
Add suggested changes

Suggestions include making pixel_to_image and image_to_pixel as
staticmethods and using np.mask rather than tuples.

mamelara commented:
Rebased to current dev and added suggested changes

oruebel reviewed (dismissed):
It would be good to add a brief docstring to the pixel_to_image and image_to_pixel methods to describe what they do, but other than that I think this looks good.

mamelara committed:
Add docstring
